pandas/io/stata.py

```diff
@@ -1141,72 +1246,136 @@ def _read_strls(self):
-            StataMissingValue objects.
-        preserve_dtypes : boolean, defaults to True
-            Preserve Stata datatypes. If False, numeric data are upcast to
-            pandas default types for foreign data (float64 or int64)
-        columns : list or None
-            Columns to retain.  Columns will be returned in the given order.
-            None returns all columns
-        order_categoricals : boolean, defaults to True
-            Flag indicating whether converted categorical data are ordered.
+        size : int, defaults to None
+            Number of lines to read.  If None, reads whole file.
 
         Returns
         -------
-        y : DataFrame instance
+        DataFrame
         """
+        if size is None:
+            size = self._chunksize
+        return self.read(nrows=size)
+
+
+    @Appender(_read_method_doc)
+    def read(self, nrows=None, convert_dates=None,
+             convert_categoricals=None, index=None,
+             convert_missing=None, preserve_dtypes=None,
+             columns=None, order_categoricals=None):
         self._missing_values = convert_missing
-        if self._data_read:
-            raise Exception("Data has already been read.")
-        self._data_read = True
 
-        if self.format_version >= 117:
+        # Handle empty file.  If reading incrementally raise
+        # StopIteration.  If reading the whole thing return an empty
+        # data frame.
+        if (self.nobs == 0) and (nrows is None):
+            self._can_read_value_labels = True
+            self._data_read = True
+            return DataFrame(columns=self.varlist)
+
+        # Handle options
+        if convert_dates is None:
+            convert_dates = self._convert_dates
+        if convert_categoricals is None:
+            convert_categoricals = self._convert_categoricals
+        if convert_missing is None:
+            convert_missing = self._convert_missing
+        if preserve_dtypes is None:
+            preserve_dtypes = self._preserve_dtypes
+        if columns is None:
+            columns = self._columns
+        if order_categoricals is None:
+            order_categoricals = self._order_categoricals
+
+        if nrows is None:
+            nrows = self.nobs
+
+        if self.format_version >= 117 and self._first_chunk:
+            self._can_read_value_labels = True
             self._read_strls()
 
+        # Setup the dtype.
+        if self._first_chunk:
+            dtype = []  # Convert struct data types to numpy data type
+            for i, typ in enumerate(self.typlist):
+                if typ in self.NUMPY_TYPE_MAP:
+                    dtype.append(('s' + str(i), self.byteorder + self.NUMPY_TYPE_MAP[typ]))
+                else:
+                    dtype.append(('s' + str(i), 'S' + str(typ)))
+            dtype = np.dtype(dtype)
+            self._dtype = dtype
+
         # Read data
-        count = self.nobs
-        dtype = []  # Convert struct data types to numpy data type
```

Comment:
side issue, are dtypes read from even say the 1st line unambiguous w.r.t. the rest of the file? IOW, these look 'given' and you are NOT inferring them? (otherwise you could have integer/float issues if you don't see any nans for example)